benchmarks/misc/end2end_cases/rasterizer.py

```diff
@@ -0,0 +1,159 @@
+        self.c2[i] = c2
+
+    @staticmethod
+    @ti.func
+    def point_in_triangle(P, A, B, C):
+        alpha = -(P.x - B.x) * (C.y - B.y) + (P.y - B.y) * (C.x - B.x)
+        alpha /= -(A.x - B.x) * (C.y - B.y) + (A.y - B.y) * (C.x - B.x)
+        beta = -(P.x - C.x) * (A.y - C.y) + (P.y - C.y) * (A.x - C.x)
+        beta /= -(B.x - C.x) * (A.y - C.y) + (B.y - C.y) * (A.x - C.x)
+        gamma = 1.0 - alpha - beta
+        result = alpha >= 0.0 and alpha <= 1.0 and beta >= 0.0 and beta <= 1.0 and gamma >= 0.0
+        return result, alpha, beta, gamma
+
+    @staticmethod
+    @ti.func
+    def bbox_intersect(A0, A1, B0, B1):
+        return (B0.x < A1.x and B0.y < A1.y and B1.x > A0.x and B1.y > A0.y)
+
+    @ti.kernel
+    def tile_culling(self):
+        for i, j in self.block_num_triangles:
+            idx = 0
+            tile_min = ti.Vector([i * self.tile_size, j * self.tile_size])
+            tile_max = ti.Vector([(i + 1) * self.tile_size,
+                                  (j + 1) * self.tile_size])
+            for t in range(self.num_triangles):
+                A, B, C = self.A[t], self.B[t], self.C[t]
+                tri_min = ti.min(A, ti.min(B, C))
+                tri_max = ti.max(A, ti.max(B, C))
+                if self.bbox_intersect(tile_min, tile_max, tri_min, tri_max):
+                    self.block_indicies[i, j, idx] = t
+                    idx = idx + 1
+            self.block_num_triangles[i, j] = idx
+
+    @ti.kernel
+    def rasterize(self):
+        for i, j in self.pixels:
+            for k in range(self.block_num_triangles[i // self.tile_size,
+                                                    j // self.tile_size]):
+                idx = self.block_indicies[i // self.tile_size,
+                                          j // self.tile_size, k]
+                A, B, C = self.A[idx], self.B[idx], self.C[idx]
+                c0, c1, c2 = self.c0[idx], self.c1[idx], self.c2[idx]
+
+                for subi, subj in ti.ndrange(self.num_spp_sqrt,
+                                             self.num_spp_sqrt):
+                    P = ti.Vector([
+                        i + (subi + 0.5) / self.num_spp_sqrt,
+                        j + (subj + 0.5) / self.num_spp_sqrt
+                    ])
+                    result, alpha, beta, gamma = self.point_in_triangle(
+                        P, A, B, C)
+
+                    if result:
+                        interpolated_color = c0 * alpha + c1 * beta + c2 * gamma
+                        self.samples[i, j, subi, subj] = interpolated_color
+
+            samples_sum = ti.Vector([0.0, 0.0, 0.0])
+            for subi, subj in ti.ndrange(self.num_spp_sqrt, self.num_spp_sqrt):
+                samples_sum += self.samples[i, j, subi, subj]
+            self.pixels[i, j] = samples_sum / self.num_samples_per_pixel
+
+
+def e2e_rasterizer(test_arch):
+    basic_repeat_times = 100
+    arch_repeat_times = 10 if test_arch == ti.cuda else 1
```

Comment:
FYI: repeat_times is used to ensure the reliability and stability of different architecture benchmarks